django/db/models/sql/query.py

```diff
@@ -1408,13 +1408,41 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
     def add_filtered_relation(self, filtered_relation, alias):
         filtered_relation.alias = alias
         lookups = dict(get_children_from_q(filtered_relation.condition))
-        for lookup in chain((filtered_relation.relation_name,), lookups):
+        relation_lookup_parts, relation_field_parts, _ = self.solve_lookup_type(filtered_relation.relation_name)
+        if relation_lookup_parts:
+            raise ValueError(
+                "FilteredRelation's relation_name doesn't support lookups "
+                "(got %r)." % filtered_relation.relation_name
+            )
+        for lookup in chain(lookups):
             lookup_parts, field_parts, _ = self.solve_lookup_type(lookup)
             shift = 2 if not lookup_parts else 1
-            if len(field_parts) > (shift + len(lookup_parts)):
+            field_path = field_parts[:-shift]
+
+            if len(field_path) <= len(relation_field_parts):
+                nested_field_path = False
+                for relation_field_part in relation_field_parts:
+                    if field_path:
+                        if field_path[0] == relation_field_part:
+                            field_path = field_path[1:]
+                        else:
+                            raise ValueError(
+                                "FilteredRelation's condition doesn't support nested "
+                                "on clauses not within the path of the relation (got %r for %r)." % (
+                                lookup, filtered_relation.relation_name)
+                            )
+                    elif not nested_field_path:
+                        nested_field_path = True
+                    else:
+                        raise ValueError(
+                            "FilteredRelation's condition doesn't support nested "
+                            "on clauses deeper than the relation (got %r for %r)." % (
+                            lookup, filtered_relation.relation_name)
+                        )
+            else:
                 raise ValueError(
```

Comment:
Probably not the right error case, it's also duplicate of the one right above it, checking inside the nested one. This can probably be removed, but I am considering if there are other errors or checks that should be made.